File: go/jzoffer/validateStackSequences_31.go
package jzoffer

import "container/list"

func validateStackSequences(pushed []int, popped []int) bool {
	stack := list.New()

	for i, j := 0, 0; i < len(pushed); i++ {
		if pushed[i] != popped[j] {
			stack.PushBack(pushed[i])
			continue
		}
		// 弹出
		j++
		// 直到 popped 和已经加入栈的不同
		for j < len(popped) && stack.Len() > 0 && stack.Back().Value.(int) == popped[j] {
			stack.Remove(stack.Back())
			j++
		}
	}

	return stack.Len() == 0
}

File: java/jzoffer/ValidateStackSequences_31.java
import java.util.Deque;
import java.util.LinkedList;

public class ValidateStackSequences_31 {

    // 模拟栈操作
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Deque<Integer> stack = new LinkedList<>();

        for (int i = 0, j = 0; i < pushed.length; i++) {
            if (pushed[i] != popped[j]) {
                stack.add(pushed[i]);
                continue;
            }
            // 等于的话 就要弹出栈了
            // 弹出 popped
            j++;

            while (stack.size() > 0 && j < popped.length && stack.getLast() == popped[j]) {
                stack.removeLast();
                j++;
            }
        }
        return stack.isEmpty();
    }
}
